src/internal_modules/roc_audio/latency_monitor.h

```diff
@@ -19,16 +19,106 @@
+    //! If the latency goes out of bounds, the session is terminated.
+    core::nanoseconds_t latency_tolerance;
+
+    //! Maximum allowed deviation of freq_coeff from 1.0.
+    //! If the scaling goes out of bounds, it is trimmed.
+    //! For example, 0.01 allows freq_coeff values in range [0.99; 1.01].
+    float scaling_tolerance;
+
+    //! Automatically tune target latency within tolarance range so as to
+    //!
+    //! increase it when:
+    //! * jitter grows
+    //! * FEC start being sent or its block length grows
+    //!
+    //! or decrease it when jitter and FEC block length allows to.
+    bool auto_tune_target_latency;
+
+    LatencyMonitorConfig()
+        : fe_enable(true)
+        , fe_profile(FreqEstimatorProfile_Responsive)
+        , fe_update_interval(5 * core::Millisecond)
+        , latency_tolerance(0)
+        , scaling_tolerance(0.005f)
+        , auto_tune_target_latency(false) {
+    }
+
+    //! Automatically deduce FreqEstimator profile from target latency.
+    void deduce_fe_profile(const core::nanoseconds_t target_latency) {
+        fe_profile = target_latency < 30 * core::Millisecond
+            // prefer responsive profile on low latencies, because gradual profile
+            // won't do it at all
+            ? FreqEstimatorProfile_Responsive
+            // prefer gradual profile for higher latencies, because it can handle
+            // higher network jitter
+            : FreqEstimatorProfile_Gradual;
+    }
+
+    //! Automatically deduce latency_tolerance from target_latency.
+    void deduce_latency_tolerance(core::nanoseconds_t target_latency) {
+        // this formula returns target_latency * N, where N starts with larger
+        // number and approaches 0.5 as target_latency grows
+        // examples:
+        //  target=1ms -> tolerance=8ms (x8)
+        //  target=10ms -> tolerance=20ms (x2)
+        //  target=200ms -> tolerance=200ms (x1)
+        //  target=2000ms -> tolerance=1444ms (x0.722)
+        if (target_latency < core::Millisecond) {
+            target_latency = core::Millisecond;
+        }
+        latency_tolerance = core::nanoseconds_t(
+            target_latency
+            * (std::log((200 * core::Millisecond) * 2) / std::log(target_latency * 2)));
+    }
+};
+
+//! Metrics of latency monitor.
+struct LatencyMonitorMetrics {
+    //! Estimated NIQ latency.
+    //! NIQ = network incoming queue.
+    //! Defines how many samples are buffered in receiver packet queue and
+    //! receiver pipeline before depacketizer (packet part of pipeline).
+    core::nanoseconds_t niq_latency;
+
+    //! Estimated E2E latency.
+    //! E2E = end-to-end.
+    //! Defines how much time passed between frame entered sender pipeline
+    //! (when it is captured) and leaved received pipeline (when it is played).
+    core::nanoseconds_t e2e_latency;
+
+    LatencyMonitorMetrics()
+        : niq_latency(0)
+        , e2e_latency(0) {
+    }
+};
+
```

Comment:
These two structs are not used in code. (Instead, structs from latency_tuner.h are used).

```suggestion
```